local Utils = {}

function Utils.getRoot(character)
	return character:FindFirstChild("HumanoidRootPart")
end

function Utils.GetCameraRelativeDirection(moveVector, cameraCFrame)
	local camLook = cameraCFrame.LookVector
	camLook = Vector3.new(camLook.X, 0, camLook.Z).Unit

	local camRight = cameraCFrame.RightVector
	camRight = Vector3.new(camRight.X, 0, camRight.Z).Unit

	local direction = (camLook * moveVector.Z) + (camRight * moveVector.X)

	if direction.Magnitude > 0 then
		return direction.Unit
	else
		return Vector3.zero
	end
end

function Utils.clamp(value, min, max)
	if value < min then
		return min
	elseif value > max then
		return max
	else
		return value
	end
end

function Utils.lerp(a, b, t)
	return a + (b - a) * t
end

function Utils.round(value)
	return math.floor(value + 0.5)
end

function Utils.sign(value)
	if value > 0 then
		return 1
	elseif value < 0 then
		return -1
	else
		return 0
	end
end

function Utils.percent(part, whole)
	if whole == 0 then
		return 0
	else
		return (part / whole) * 100
	end
end


function Utils.deepCopy(original)
	local copy
	if type(original) == 'table' then
		copy = {}
		for key, value in next, original, nil do
			copy[Utils.deepCopy(key)] = Utils.deepCopy(value)
		end
		setmetatable(copy, Utils.deepCopy(getmetatable(original)))
	else
		copy = original
	end
	return copy
end

return Utils